jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/BufferedContentSink.java

```diff
@@ -0,0 +1,182 @@
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.io.content;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+
+import org.eclipse.jetty.io.ByteBufferAggregator;
+import org.eclipse.jetty.io.ByteBufferPool;
+import org.eclipse.jetty.io.Content;
+import org.eclipse.jetty.io.RetainableByteBuffer;
+import org.eclipse.jetty.util.BufferUtil;
+import org.eclipse.jetty.util.Callback;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * <p>A {@link Content.Sink} backed by another {@link Content.Sink}.
+ * Any content written to this {@link Content.Sink} is buffered,
+ * then written to the delegate using
+ * {@link Content.Sink#write(boolean, ByteBuffer, Callback)}. </p>
+ */
+public class BufferedContentSink implements Content.Sink
+{
+    private static final Logger LOG = LoggerFactory.getLogger(BufferedContentSink.class);
+
+    private static final int START_BUFFER_SIZE = 1024;
+
+    private final Content.Sink _delegate;
+    private final ByteBufferPool _bufferPool;
+    private final boolean _direct;
+    private final int _maxBufferSize;
+    private ByteBufferAggregator _aggregator;
+    private boolean _firstWrite = true;
+    private boolean _lastWritten;
+
+    public BufferedContentSink(Content.Sink delegate, ByteBufferPool bufferPool, boolean direct, int maxBufferSize)
+    {
+        if (maxBufferSize <= 0)
+            throw new IllegalArgumentException("maxBufferSize must be > 0, was: " + maxBufferSize);
+        _delegate = delegate;
+        _bufferPool = (bufferPool == null) ? new ByteBufferPool.NonPooling() : bufferPool;
+        _direct = direct;
+        _maxBufferSize = maxBufferSize;
+    }
+
+    @Override
+    public void write(boolean last, ByteBuffer byteBuffer, Callback callback)
+    {
+        if (LOG.isDebugEnabled())
+            LOG.debug("writing last={} {}", last, BufferUtil.toDetailString(byteBuffer));
+
+        if (_lastWritten)
+        {
+            callback.failed(new IOException("complete"));
+            return;
+        }
+        if (_firstWrite)
+        {
+            _firstWrite = false;
+            if (last)
+            {
+                // No need to buffer if this is both the first and the last write.
+                _lastWritten = true;
+                _delegate.write(true, byteBuffer, callback);
+                return;
+            }
+            _aggregator = new ByteBufferAggregator(_bufferPool, _direct, Math.min(START_BUFFER_SIZE, _maxBufferSize), _maxBufferSize);
+        }
+        _lastWritten |= last;
+
+        ByteBuffer current = byteBuffer != null ? byteBuffer : BufferUtil.EMPTY_BUFFER;
+        if (current.remaining() <= _maxBufferSize)
```

Comment:
Rather than testing against maxBufferSize,  it would be good to test against a heuristically determined commit size.    

This probably argues against my point above regarding delaying the creation of the aggregator, as it is probably the right place to put such logic.
Only once you have such logic an this class be used in the `HttpOutput` implementations.l